### PR TITLE
fix the value of default shells to use /bin and not /usr/bin

### PR DESCRIPTION
**Pull request by eli-schwartz**
Partially reverts commit b0d3095fd6cc1791a38f57a1982116b4475244ba.

While it is generally worthwhile for systemd to drop split-usr support, these options are NOT about split-usr support. The universal location of POSIX sh is always /bin/sh. Bash is pretty reasonably standardized there too.

This happens irrespective of /bin being a symlink to /usr/bin. Ramifications of this change include things like:

- portably running shell scripts that might run very nearly anywhere
- /etc/shells support

For standardization and compatibility reasons, these commands with these paths need to be consistently found on any system, and thus distros make sure this works, although even on split-usr systems /usr/bin/bash may be a symlink to /bin/bash.

Embedding the *access path* of bash as /usr/bin/bash in systemd, for example in libnss_systemd.so, means that login shells must agree with systemd on how they invoke the shell. End result: users fail to login because of access violations.

This cannot be fixed by "fixing PAM" because PAM does not follow symlinks by design: one example is that it needs to treat rbash as different from bash.

Fixes: https://bugs.gentoo.org/919749

### CODE REVIEW

**Comment by YHNdnzj** [comment]
These are well-defined meson options. Why can't distros override them on their own?

Also, generally after the drop of split-usr support, distros should put effort into ensuring that `/usr/bin/` and `/bin/` are treated equally wrt `/etc/shells` and friends. For example, Arch does both: https://gitlab.archlinux.org/archlinux/packaging/packages/filesystem/-/blob/main/shells?ref_type=heads

So I'm not sure changing this upstream...

**Comment by YHNdnzj** [comment]
I mean, users should expect the same behavior when they do `chsh -s /usr/bin/bash` and `chsh -s /bin/bash` too, right?

**Comment by eli-schwartz** [comment]
> These are well-defined meson options. Why can't distros override them on their own?

The current value is ***always*** wrong. Not just sometimes wrong. Not even "most of the time" wrong.

Always, always, always wrong.

Why not fix it? Why leave footguns lying around for unsuspecting people that give them a sour taste in their mouth when they try to use systemd and it's randomly broken?

> Also, generally after the drop of split-usr support, distros should put effort into ensuring that `/usr/bin/` and `/bin/` are treated equally wrt `/etc/shells` and friends. For example, Arch does both:

Because this is a breaking change in systemd which was ***unintentional*** and broke users of merged-usr systems.

Because it's easier to fix something in one place than in dozens of places.

Or just because it is semantically correct!

> I mean, users should expect the same behavior when they do `chsh -s /usr/bin/bash` and `chsh -s /bin/bash` too, right?

Perhaps if I actually saw people chsh'ing to the paths in /usr, but I do not...

Either way this is obviously something the user manually did and can manually know about and manually debug. Having the system silently break because someone did all the right things and used merged-usr as the changelog told them to do, and concluded that updating systemd was safe, is pretty broken.

Also I don't actually know why someone using chsh would want to type longer paths out by hand but ymmv...

**Comment by thesamesam** [comment]
In addition to Eli's arguments:

> So I'm not sure about changing this upstream...

I don't actually see a reason not to. One distro has done this, after it was open on their tracker for _10_ years (https://bugs.archlinux.org/task/33677) (I don't blame them, there's a bunch of issues at play, just saying it's not trivial), and they did so saying _"for now_".

We also had real users broken by the original change Eli is fixing who did everything right. The systemd way tends to be "just works". Given that nobody seems to have acknowledged the `/etc/shells` problem, it comes across like it was unintentional/a bug, rather than some move designed to push people to migrate - as you suggest people should do.

If it _were_ intentional, it would've been in the NEWS and such, no?

**Comment by YHNdnzj** [comment]
> The current value is always wrong. Not just sometimes wrong. Not even "most of the time" wrong.
> 
> Always, always, always wrong.

At least Arch has switched the path of root's default shell to `/usr/bin/bash`, so I'm not convinced at all... https://gitlab.archlinux.org/archlinux/packaging/packages/filesystem/-/commit/adf12b1575dbb3cbb90a02bdb3b5a00d4d8effd4

**Comment by awilfox** [comment]
Setting aside `/usr/bin/bash` (which I agree should be reverted to `/bin/bash` as well) for a moment:

`/bin/sh` has a special meaning on all POSIX and POSIX-like systems, and that is to be the base shell.  glibc hardcodes `/bin/sh` and `/bin/csh` to be the only valid shells for `getusershell` [if `/etc/shells` does not exist](https://sourceware.org/git/?p=glibc.git;a=blob;f=misc/getusershell.c;h=4221095dca743dfa5067637f1aad4651bd5cf279;hb=HEAD#l113).  Programs exist which [test whether argv[0] is "/bin/sh"](https://sources.debian.org/src/libpod/4.9.3+ds1-1/pkg/specgen/generate/container_create.go/?hl=404#L404) (which would fail if it is /usr/bin/sh).  Trust me, this is not a change you want to make.

**Comment by YHNdnzj** [comment]
> We also had real users broken by the original change Eli is fixing who did everything right. The systemd way tends to be "just works". Given that nobody seems to have acknowledged the /etc/shells problem, it comes across like it was unintentional/a bug, rather than some move designed to push people to migrate - as you suggest people should do. That is the opposite of "just works".
> 
> If it were intentional, it would've been in the NEWS and such, no?

Your points do make more sense I think. I'm still not convinced that using `/usr/bin/bash` is anything *wrong*, and eventually this should be made available on all usr-merged systems. But if Arch is the only one that is currently unaffected, and it's unintentional, maybe we should revert this, at least for now.

So LGTM now.

**Comment by poettering** [comment]
we have dropped support for split /usr/ though in v255, hence why does this even matter?

**Comment by mbiebl** [comment]
> Your points do make more sense I think. I'm still not convinced that using `/usr/bin/bash` is anything _wrong_, and eventually this should be made available on all usr-merged systems. But if Arch is the only one that is currently unaffected, and it's unintentional, maybe we should revert this, at least for now.

Is this about /etc/shells having an entry for `/bin/bash` but software that consults `/etc/shells` using `/usr/bin/bash`?
If so, this can be worked around by listing both paths in `/etc/shells`. This is what Fedora does and we decided to do the same in Debian/Ubuntu (even if the duplication is a bit ugly).

**Comment by YHNdnzj** [comment]
> > Your points do make more sense I think. I'm still not convinced that using `/usr/bin/bash` is anything _wrong_, and eventually this should be made available on all usr-merged systems. But if Arch is the only one that is currently unaffected, and it's unintentional, maybe we should revert this, at least for now.
> 
> Is this about /etc/shells having an entry for `/bin/bash` but software that consults `/etc/shells` using `/usr/bin/bash` failing as the paths don't match? If so, this can be worked around by listing both paths in `/etc/shells`. This is what Fedora does and we decided to do the same in Debian/Ubuntu (even if the duplication is a bit ugly).

Ah, so every major distro except Gentoo is fine? Then I guess it's better for Gentoo to fix this themselves (I only checked Arch).

**Comment by bluca** [comment]
> > Your points do make more sense I think. I'm still not convinced that using `/usr/bin/bash` is anything _wrong_, and eventually this should be made available on all usr-merged systems. But if Arch is the only one that is currently unaffected, and it's unintentional, maybe we should revert this, at least for now.
> 
> Is this about /etc/shells having an entry for `/bin/bash` but software that consults `/etc/shells` using `/usr/bin/bash` failing as the paths don't match? If so, this can be worked around by listing both paths in `/etc/shells`. This is what Fedora does and we decided to do the same in Debian/Ubuntu (even if the duplication is a bit ugly).

Yes indeed, you can just list both and it works fine. This is the intended default, and it's right that it is the modern thing. It's an option exactly to allow overriding it for older legacy systems.

**Comment by poettering** [comment]
So even though I think gentoo is just causing this issue for themselves, I think the proposed fix might be acceptable. I do accept that /bin/sh is kinda API and there's little reason to purposefully deviate from it given it's fully equivalent to /usr/bin/sh on all systems we support...

I mean, paths.h defines 

```
#define _PATH_BSHELL    "/bin/sh"
```

thus one could claim it's not just "kinda API" but "actually API". But then again, paths.h defines all kinds of incorrect and out-of-date garbage...

**Comment by bluca** [comment]
> So even though I think gentoo is just causing this issue for themselves, I think the proposed fix might be acceptable. I do accept that /bin/sh is kinda API and there's little reason to purposefully deviate from it given it's fully equivalent to /usr/bin/sh on all systems we support...

I don't think we should change the default for the sake of a single distro. The current default works fine in every other major distribution. As mentioned, fixing /etc/shells to list both paths is trivial and what everybody else already does. Gentoo can either do the same thing, or override the default setting with the provided options.

**Comment by keszybz** [comment]
At least Arch and Fedora work just fine with the new paths.

> I mean, users should expect the same behavior when they do `chsh -s /usr/bin/bash` and `chsh -s /bin/bash` too, right?

Yeah. We switched to the new paths, and the only thing that makes sense is to accept the new paths. I think it also makes sense to keep accepting the old paths for compatibility. Any system which rejects the new path but accepts the old path is severely broken. That is the real *footgun* and a way to annoy users.

> `/bin/sh` has a special meaning on all POSIX and POSIX-like systems, and that is to be the base shell. glibc hardcodes `/bin/sh` and `/bin/csh` to be the only valid shells for `getusershell` [if `/etc/shells` does not exist](https://sourceware.org/git/?p=glibc.git;a=blob;f=misc/getusershell.c;h=4221095dca743dfa5067637f1aad4651bd5cf279;hb=HEAD#l113).

This is something to fix in glibc. At this point, `/usr/bin/sh` is **the** path, while `/bin/sh` is a compat symlink that is very likely to exist, but conceivably might be missing on a usr-only system sometime in early boot.

In summary: I'm not opposed to doing this, but I'd like to see a good *technical* justification why it's better to have `/bin/sh` instead of `/usr/bin/sh`. Just saying "it's wrong" is not a justification.

**Comment by Jannik2099** [comment]
> Ah, gentoo being bitten by their franken-/usr-merge, eh?

This is unrelated to our usr-merge, which, by the way, produces the same directory layout as Fedora.

> The current default works fine in every other major distribution.

I would argue the current default does not work at all since it does not match glibc behaviour, and only works in said distros because everyone had to add the workaround in `/etc/shells` - with a "stock" `/etc/shells` that only lists the true, non-symlinked paths, the current defaults do NOT work.

It's also reasonable to think that other tooling may read `/etc/shells` without following symlinks. Less reasonably, some "minimalist" may run a system that lacks `/etc/shells`, which used to work prior to this change. Is `/etc/shells` now a systemd requirement?

> This is something to fix in glibc. At this point, /usr/bin/sh is the path, while /bin/sh is a compat symlink that is very likely to exist, but conceivably might be missing on a usr-only system sometime in early boot.

`/bin/sh` is indeed an outdated glibc-ism, but I think getting this change through glibc will be *significantly* slower than fixing it on this end.

As it is, systemd and glibc contradict each other, and POSIX does not specify the location of `sh`

On that note, does the systemd usr-merge specify that the paths in `/etc/passwd` should be updated? Fedora and Gentoo both still use `/bin/bash` there.

**Comment by bluca** [comment]
> with a "stock" /etc/shells

No such thing. It's a config file, you list what you want to list in it.

> Is /etc/shells now a systemd requirement?

No, it's a requirement for those pam modules that use it

**Comment by mbiebl** [comment]
> So this is about /etc/shells being out of date in gentoo, and then using pam_shells? frankly, pam_shells is such a useless concept. I understand gentoo is a conservative distro, but why do you even still have that in your PAM stacks? in fedora it's _long_ gone...
> 
> Not sure what's more useless pam_shells or pam_securetty... Both propagate a fake concept of security, that is cargo culted around by some unix truthers for no reason but to create problems for themselves.

Apparently, even PolicyKit consults /etc/shells: https://lists.debian.org/debian-devel/2024/02/msg00188.html

**Comment by floppym** [comment]
> Ah, gentoo being bitten by their franken-/usr-merge, eh?

Please try not to make such uninformed attacks. It just makes you look like a bully.

**Comment by poettering** [comment]
> Please try not to make such uninformed attacks. It just makes you look like a bully.

Now i am curious. I thought that Gentoo was trying to do the /usr/ merge via symlinking individual files between /usr/bin/ and /bin/, rather than the whole dir? Was that approach abandoned in the end? Because if not, and this is still the approach Gentoo chases, then I would still call this frankenusrmerge, since it really only gives you the disadvantages of the merge without giving the the advantages of it…

**Comment by poettering** [comment]
> (We ended up clarifying it but I think it got lost in the very busy previous PR: [#27999 (comment)](https://github.com/systemd/systemd/pull/27999#discussion_r1226307120). But yeah, we're doing the proper Fedora-style usr-merge. No cheesy individual symlinks.)

Ah, interesting. Seems you educated me on this before, but it didn't stick. Let's see if I can do better this time.

**Comment by eli-schwartz** [comment]
Apologies for the lateness of the reply. I was feeling quite sick this past week.

> I don't think we should change the default for the sake of a single distro. The current default works fine in every other major distribution.

It's not changing the default.

It's reverting *your* breaking change to the default that was not even announced.

> This is something to fix in glibc. At this point, `/usr/bin/sh` is the path, while `/bin/sh` is a compat symlink that is very likely to exist, but conceivably might be missing on a usr-only system sometime in early boot.
>
> In summary: I'm not opposed to doing this, but I'd like to see a good technical justification why it's better to have `/bin/sh` instead of `/usr/bin/sh`. Just saying "it's wrong" is not a justification.

"It breaks the glibc Linux API" isn't a good enough reason?

But I certainly don't expect glibc to humor your belief that /usr/bin/sh is "the path".

/bin/sh is the glibc API because glibc:
- aligns with the POSIX standard, as well as, crucially, the userland API of Unix systems as a whole, and that means "/bin/sh, no quibbling"
- supports other init systems than systemd, including those where usr merge is optional.

The Unix standard is that shell scripts containing the shebang `#!/usr/bin/sh` are factually incorrect, and those that contain the shebang `#!/bin/sh` are correct.

Once again, I say: this has nothing to do with the usr merge. Regardless of distros such as Gentoo implementing the usr merge in the straightforward and standard manner that systemd wants done (despite repeated FUD by systemd developers that turned out to be factually incorrect), the correct and canonical location for these programs is not "the physical location on disk" but "the API name".

By the way:
> but conceivably might be missing on a usr-only system sometime in early boot.

I'm very confused by this statement. I'm assuming this is talking about the "hermetic read-only /usr mounted to a freshly initialized rootfs" scenario. My understanding was that this requires creating a blank rootfs, then initializing a basic FHS structure including the mountpoint for /usr itself, as well as a mountpoint for home directories, probably creating the /root skeleton, and so on.

As part of the same transaction creating /usr, one would naturally create a /bin and /lib symlink pointing to the (not yet mounted) /usr hierarchy.

While indeed you might be missing /bin/sh and /bin/bash sometime during early boot, I would say that the only time this is possible is before /usr/bin/sh exists either, since /usr hasn't been mounted and you're still inside the initramfs.

I wouldn't like to bet on it being a good idea to be missing those symlinks during early boot anyway, since arbitrary programs that you run during early boot, if they use shell at all, will have shebangs with `#!/bin/sh` and thus break.

You'd also, naturally, break glibc APIs for running subprocesses, for example the `system()` function, since glibc would expect /bin/sh to exist.

Yes, of course, `system()` is a bit gunky compared to the `exec()` family of functions, but still -- are you REALLY going to argue that the systemd usrmerge of 2023 has dropped support for "software that uses `system()`"?

I think not. I think *absolutely* not. I think that the canonical name for the `sh` shell remains "access it from the /bin location and allow the filesystem to resolve the realpath".

This whole discussion about "the new name" is a red herring. There is no new name.

**Comment by eli-schwartz** [comment]
On the general applicability of "just add additional shells to /etc/shells", I fear this is missing the point. I wouldn't *mind* adding it, though it's not actually up to me and for openrc users that share a base layout package with systemd it could be considered "bad and wrong", so idk.

But even in the situation where it is added, I still consider it wrong and dangerous to default to /usr/bin/bash.

Traditionally, whatever is configured to work on your system is good enough. If you want a non-default shell, configure /etc/shells and all is good. But systemd is *specifically* trying to upset the status quo. That's the whole point of systemd-homed. If systemd succeeds and other people start implementing homed for, say, Alpine systems, then we will end up with people creating homed user accounts on Arch or Fedora, using them successfully for a while, and then trying to log in on Alpine and discovering to their dismay and shock that Alpine doesn't use systemd and doesn't care about split-usr, and /bin/bash exists if you install bash, but /usr/bin/bash won't ever exist.

For this reason, we should be looking to the future and strictly rejecting any and all attempts to create new user accounts that describe the shell using a non-API filename in /usr.

There is zero risk or harm in using /bin/bash, and a great deal of risk in using /usr/bin/bash.

Again, this is about API filenames, not usrmerge.

**Comment by bluca** [comment]
> (despite repeated FUD by systemd developers that turned out to be factually incorrect)

This was already clarified as an honest misunderstanding, already corrected. Please, let's not drag it any further, there's no need.

> If systemd succeeds and other people start implementing homed for, say, Alpine systems

I am quite sure we can confidently say there are exactly zero chances of anything like that happening, at any point, in this universe or the next.

> There is zero risk or harm in using /bin/bash, and a great deal of risk in using /usr/bin/bash.

The thing is, the two are entirely equivalent from our point of view. One is a legacy link to the other, and either can be used, so we use it.

**Comment by awilfox** [comment]
> > If systemd succeeds and other people start implementing homed for, say, Alpine systems
> 
> I am quite sure we can confidently say there are exactly zero chances of anything like that happening, at any point, in this universe or the next.

I wouldn't be so sure.  Chimera has sysusers and tmpfiles as well as systemd-boot.  I'm personally planning on writing a timedate1 implementation for non-sd systems later this year.  If homed starts having real adoption and can be shown to solve real problems, I don't know why alternative implementations (or just building it standalone) couldn't happen.


> > There is zero risk or harm in using /bin/bash, and a great deal of risk in using /usr/bin/bash.
> 
> The thing is, the two are entirely equivalent from our point of view. One is a legacy link to the other, and either can be used, so we use it.

If the two are entirely equivalent from your point of view, it should be easy then to revert this change and use the links that are guaranteed to exist on all systems.  This makes the profiles more portable to other Linux systems.  There is no upside to changing this, not even binary size reduction.

**Comment by poettering** [comment]
> I wouldn't be so sure. Chimera has sysusers and tmpfiles as well as systemd-boot. I'm personally planning on writing a timedate1 implementation for non-sd systems later this year. If homed starts having real adoption and can be shown to solve real problems, I don't know why alternative implementations (or just building it standalone) couldn't happen.

What is chimera, and are you saying they are reimplementing sysusers and tmpfiles and sd-boot? why tf would they do that, sysusers/tmpfiles/sd-boot are all open source?

but dude, what can i say, if you indeed want to weaken your position in a discussion with us, it's generally a fantastic good idea to put the angle "alternative to systemd" front and center of your discussion points. I mean, do as many alternatives as you want, but if you are asking us to be particularly accomodating for that, then oh man, you are so very mistaken.

**Comment by poettering** [comment]
so i still think we should just stick to the /bin/sh and /bin/bash paths simply because people do dumb string comparisons on these strings (that is what /etc/shells after all is used for, as dumb as the concept might be) and I see no reason to trip those cases up, because nothing is gained by doing so.

let's not forget that we ourselves hardcode /bin/sh all over the place in our codebase still, for a reason... 

But I also don't care too much. And I don't understand why gentoo doesn't simply use the meson options and is done with it...

**Comment by mbiebl** [comment]
I think Gentoo should allow `/usr/bin/bash` to be a valid login shell, i.e. list it in `/etc/shells`, if they apply usrmerge.

**Comment by YHNdnzj** [comment]
Yeah, as mentioned by @keszybz and @mbiebl already, refusing new paths after usr-merge is definitely wrong. It's natural for users to assume `/usr/bin/bash` and `/bin/bash` are equivalents after they see the `/bin/` -> `/usr/bin/` symlink. There's simply no reason to make their life hard.

**Comment by awilfox** [comment]
> What is chimera, and are you saying they are reimplementing sysusers and tmpfiles and sd-boot? why tf would they do that, sysusers/tmpfiles/sd-boot are all open source?

No, none of it is reimplemented, it is all using systemd upstream code.  It doesn't need to be reimplemented.  The systemd code works great.

> but dude, what can i say, if you indeed want to weaken your position in a discussion with us, it's generally a fantastic good idea to put the angle "alternative to systemd" front and center of your discussion points. I mean, do as many alternatives as you want, but if you are asking us to be particularly accomodating for that, then oh man, you are so very mistaken.

I mentioned building homed standalone, similar to many other systemd tools which are nice to run everywhere (like sd-boot, like udev).  I am hopeful that there would be no reason to write an alternative, because systemd upstream can just be used.  I personally think homed is a fantastic idea and look forward to it becoming more widely used.

**Comment by keszybz** [comment]
OK, I guess that there's general agreement that either this doesn't make much of a difference, or is beneficial. I'll go ahead and merge.
